### PR TITLE
Propagate database quarantine before dropping it

### PR DESCRIPTION
**Pull request by elprans**
Even though there may not be active connections to a database being
dropped, adjacent servers may still be keeping warm Postgres connections in
the pool, so use signaling to trigger remote pruning.

### CODE REVIEW

File: edb/server/server.py [hunk 1]
```diff
@@ -1262,6 +1305,18 @@ async def _signal_sysevent(self, event, **kwargs):
...
 
+    def _on_remote_database_quarantine(self, dbname):
+        if not self._accept_new_tasks:
+            return
+
+        # Block new connections to the database.
+        self._block_new_connections.add(dbname)
+
+        async def task():
+            await self._pg_pool.prune_inactive_connections(dbname)
```
Comment (msullivan) [inline edb/server/server.py:1316]: Can you just pass self._pg_pool.prune_inactive_connections(dbname) to `create_task`?

File: tests/test_server_proto.py [hunk 2]
```diff
@@ -2212,6 +2212,10 @@ async def test_server_adjacent_database_propagation(self):
                     await con2.query("select 1")
                     await con2.aclose()
 
+            await self.con.execute('''
+                DROP DATABASE test_db_prop;
+            ''')
```
Comment (msullivan) [inline tests/test_server_proto.py:2217]: Should we test *recreating* a database with the same name, after a drop?

Comment (msullivan) [reply tests/test_server_proto.py:2217]: (And connecting to the new one, I guess)

**Review by msullivan — APPROVED** [review]
Looks good, though a few questions

File: edb/common/retryloop.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+class RetryIteration:
+
+    def __init__(self, loop: RetryLoop) -> None:
+        self._loop = loop
+
+    async def __aenter__(self) -> RetryIteration:
+        return self
+
+    async def __aexit__(
```
Comment (1st1) [inline edb/common/retryloop.py:95]: Around here I started to realize that this might be my code... :)

File: tests/test_server_proto.py (hunk 2 above)
Comment (elprans) [reply tests/test_server_proto.py:2232]: Done.

File: edb/server/server.py [hunk 4]
```diff
@@ -1089,21 +1093,60 @@ async def _on_before_create_db_from_template(
...
         else:
-            # If, however, there are no open EdgeDB connections, prune
-            # all non-active postgres connection to the `dbname` DB.
+            # Block new connections to the database.
+            self._block_new_connections.add(dbname)
+
+            # Prune our inactive connections.
             await self._pg_pool.prune_inactive_connections(dbname)
 
+            # Signal adjacent servers to prune their connections to this
```
Comment (1st1) [inline edb/server/server.py:1109]: Can you add a fast path here awaiting on `_pg_ensure_database_not_connected`?

**Review by 1st1 — APPROVED** [review]

**Pull request merged**